lib/matplotlib/_constrained_layout.py

```diff
@@ -83,181 +74,357 @@ def do_constrained_layout(fig, renderer, h_pad, w_pad,
-
-        # OK, the above lines up ax._poslayoutbox with ax._layoutbox
-        # now we need to
-        #   1) arrange the subplotspecs.  We do it at this level because
-        #      the subplotspecs are meant to contain other dependent axes
-        #      like colorbars or legends.
-        #   2) line up the right and left side of the ax._poslayoutbox
-        #      that have the same subplotspec maxes.
-
-        if fig._layoutbox.constrained_layout_called < 1:
-            # arrange the subplotspecs...  This is all done relative to each
-            # other.  Some subplotspecs contain axes, and others contain
-            # gridspecs the ones that contain gridspecs are a set proportion
-            # of their parent gridspec.  The ones that contain axes are
-            # not so constrained.
-            figlb = fig._layoutbox
-            for child in figlb.children:
-                if child._is_gridspec_layoutbox():
-                    # This routine makes all the subplot spec containers
-                    # have the correct arrangement.  It just stacks the
-                    # subplot layoutboxes in the correct order...
-                    _arrange_subplotspecs(child, hspace=hspace, wspace=wspace)
-
-            for gs in gss:
-                _align_spines(fig, gs)
-
-        fig._layoutbox.constrained_layout_called += 1
-        fig._layoutbox.update_variables()
-
-        # check if any axes collapsed to zero.  If not, don't change positions:
-        if _axes_all_finite_sized(fig):
-            # Now set the position of the axes...
-            for ax in fig.axes:
-                if ax._layoutbox is not None:
-                    newpos = ax._poslayoutbox.get_rect()
-                    # Now set the new position.
-                    # ax.set_position will zero out the layout for
-                    # this axis, allowing users to hard-code the position,
-                    # so this does the same w/o zeroing layout.
-                    ax._set_position(newpos, which='original')
-            if do_suptitle:
-                newpos = suptitle._layoutbox.get_rect()
-                suptitle.set_y(1.0 - h_pad)
-            else:
-                if suptitle is not None and suptitle._layoutbox is not None:
-                    suptitle._layoutbox.edit_height(0)
+
+        # make margins for all the axes and subpanels in the
+        # figure.  Add margins for colorbars...
+        _make_layout_margins(fig, renderer, h_pad=h_pad, w_pad=w_pad,
+                             hspace=hspace, wspace=wspace)
+
+        _make_margin_suptitles(fig, renderer, h_pad=h_pad, w_pad=w_pad)
+
+        # if a layout is such that a columns (or rows) margin has no
+        # constraints, we need to make all such instances in the grid
+        # match in margin size.
+        _match_submerged_margins(fig)
+
+        # update all the variables in the layout.
+        fig._layoutgrid.update_variables()
+
+        if _check_ok(fig):
```

Comment:
1) check_ok could be renamed to something more descriptive e.g. _check_no_collapsed_axes
2) it's not clear why you want to skip repositioning if there are collapsed axes?

Reply:
1) Done
2) It means that no solution was found so its better not to reposition based on that because it can look bad.